RecoHI/HiJetAlgos/plugins/HiPuRhoProducer.cc

```diff
@@ -0,0 +1,615 @@
+      vntow[vi] = nt;
+    }
+
+    LogDebug("PileUpSubtractor") << " ieta: " << it << " number of towers: " << nt << " e1: " << e1 << " e2: " << e2
+                                 << "\n";
+
+    if (nt > 0) {
+      if (postOrphan_) {
+        if (nt > (int)minimumTowersFraction_ * (gt.second)) {
+          emean_[it] = e1 / (double)nt;
+          double eee = e2 / (double)nt - e1 * e1 / (double)(nt * nt);
+          if (eee < 0.)
+            eee = 0.;
+          esigma_[it] = nSigmaPU_ * sqrt(eee);
+
+          uint32_t numToCheck = std::min(int(eTop4_[it].size()), nt - (int)minimumTowersFraction_ * (gt.second));
+
+          for (unsigned int lI = 0; lI < numToCheck; ++lI) {
+            if (eTop4_[it][lI] >= emean_[it] + towSigmaCut_ * esigma_[it] && towSigmaCut_ > 0) {
+              e1 -= eTop4_[it][lI];
+              nt -= 1;
+            } else
+              break;
+          }
+
+          if (e1 < .000000001)
+            e1 = 0;
+        }
+      }
+
+      emean_[it] = e1 / (double)nt;
+      double eee = e2 / nt - e1 * e1 / (nt * nt);
+      if (eee < 0.)
+        eee = 0.;
+      esigma_[it] = nSigmaPU_ * sqrt(eee);
+
+      double etaWidth = etaedge[abs(it)] - etaedge[abs(it) - 1];
+      if (etaWidth < 0)
+        etaWidth *= -1.;
+
+      int sign = (it < 0) ? -1 : 1;
+
+      if (sign * etaedge[abs(it)] < sign * etaedge[abs(it) - 1]) {
+        etaEdgeLow_.push_back(sign * etaedge[abs(it)]);
+        etaEdgeHi_.push_back(sign * etaedge[abs(it) - 1]);
+      } else {
+        etaEdgeHi_.push_back(sign * etaedge[abs(it)]);
+        etaEdgeLow_.push_back(sign * etaedge[abs(it) - 1]);
+      }
+
+      if (vi < nEtaTow_) {
+        vmean1[vi] = emean_[it];
+        vrho1[vi] = emean_[it] / (etaWidth * (2. * M_PI / (double)vngeom[vi]));
+        rho_.push_back(vrho1[vi]);
+        rhoM_.push_back(0);
+        vrms1[vi] = esigma_[it];
+        if (vngeom[vi] == vntow[vi]) {
+          vmean0[vi] = emean_[it];
+          vrho0[vi] = emean_[it] / (etaWidth * (2. * M_PI / (double)vngeom[vi]));
+          vrms0[vi] = esigma_[it];
+        }
+        rhoExtra_.push_back(vrho0[vi]);
+        nTow_.push_back(vntow[vi]);
+      }
+    } else {
+      emean_[it] = 0.;
+      esigma_[it] = 0.;
+    }
+    LogDebug("PileUpSubtractor") << " ieta: " << it << " Pedestals: " << emean_[it] << "  " << esigma_[it] << "\n";
+  }
+
+  postOrphan_ = false;
+}
+
+void HiPuRhoProducer::subtractPedestal(std::vector<fastjet::PseudoJet>& coll) {
+  LogDebug("PileUpSubtractor") << "The subtractor subtracting pedestals...\n";
+
+  std::vector<fastjet::PseudoJet> newcoll;
+  for (auto& input_object : coll) {
+    int index = input_object.user_index();
+    reco::Candidate const* itow = inputs_[index];
+
+    const CaloTower* ctc = dynamic_cast<const CaloTower*>(itow);
+    int it = ctc->id().ieta();
+
+    double original_Et = itow->et();
+    double etnew = original_Et - (emean_.find(it))->second - (esigma_.find(it))->second;
```

Comment:
```suggestion
    double etnew = original_Et - emean_.at(it).second - esigma_.at(it).second;
```

Reply:
Not sure how to fix `error: request for member 'second' in '((HiPuRhoProducer*)this)->HiPuRhoProducer::emean_.std::map<int, double>::at(it)', which is of non-class type 'std::map<int, double>::mapped_type' {aka 'double'}`

Reply:
ah, sorry, I forgot to drop .second